newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/index.js

```diff
@@ -0,0 +1,565 @@
+// @flow
+import * as React from 'react';
+import { Line, Column } from '../../../UI/Grid';
+import Checkbox from 'material-ui/Checkbox';
+import SelectField from 'material-ui/SelectField';
+import MenuItem from 'material-ui/MenuItem';
+import SemiControlledTextField from '../../../UI/SemiControlledTextField';
+import ImagePreview from '../../../ResourcesList/ResourcePreview/ImagePreview';
+import ResourceSelector from '../../../ResourcesList/ResourceSelector';
+import ResourcesLoader from '../../../ResourcesLoader';
+import ShapePreview from './ShapePreview.js';
+import PolygonEditor from './PolygonEditor.js';
+import {
+  type ResourceSource,
+  type ChooseResourceFunction,
+} from '../../../ResourcesList/ResourceSource.flow';
+import { type ResourceExternalEditor } from '../../../ResourcesList/ResourceExternalEditor.flow';
+
+type Props = {|
+  behavior: Object,
+  project: gdProject,
+  resourceSources: Array<ResourceSource>,
+  onChooseResource: ChooseResourceFunction,
+  resourceExternalEditors: Array<ResourceExternalEditor>,
+|};
+
+type State = {|
+  image: string,
+  imageWidth: number,
+  imageHeight: number,
+|};
+
+export default class Physics2Editor extends React.Component<Props, State> {
+  resourcesLoader: typeof ResourcesLoader;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.resourcesLoader = ResourcesLoader;
+
+    this.state = {
+      image: '',
+      imageWidth: 9,
+      imageHeight: 0,
+    };
+  }
+
+  _setImageSize = (width: number, height: number) => {
+    this.setState({
+      imageWidth: width,
+      imageHeight: height,
+    });
+  };
+
+  _isPolygonConvex(properties: Object) {
+    const vertices = JSON.parse(properties.get('vertices').getValue());
+    // Get edges
+    var edges = [];
+    var v1 = null;
+    var v2 = null;
+    for (var i = 0; i < vertices.length; i++) {
+      v1 = vertices[i];
+      if (i + 1 >= vertices.length) v2 = vertices[0];
+      else v2 = vertices[i + 1];
+      edges.push({ x: v2.x - v1.x, y: v2.y - v1.y });
+    }
+    // Check convexity
+    if (edges.length < 3) return false;
+
+    const zProductIsPositive =
+      edges[0].x * edges[0 + 1].y - edges[0].y * edges[0 + 1].x > 0;
+
+    for (i = 1; i < edges.length - 1; ++i) {
+      var zCrossProduct =
+        edges[i].x * edges[i + 1].y - edges[i].y * edges[i + 1].x;
+      if (zCrossProduct > 0 !== zProductIsPositive) return false;
+    }
+
+    var lastZCrossProduct =
+      edges[edges.length - 1].x * edges[0].y -
+      edges[edges.length - 1].y * edges[0].x;
+    if (lastZCrossProduct > 0 !== zProductIsPositive) return false;
+
+    return true;
+  }
+
+  _isBitEnabled(bitsValue: number, i: number) {
+    return !!(bitsValue & (1 << i));
+  }
+
+  _enableBit(bitsValue: number, i: number, enable: boolean) {
+    if (enable) bitsValue |= 1 << i;
+    else bitsValue &= ~(1 << i);
+    return bitsValue;
+  }
+
+  renderBitProperty(
+    properties: Object,
+    isLayer: boolean,

```

Comment:
Rather than passing isLayer, it would be a bit more extensible if you pass "propertyName" :)

Reply:
Yeah, don't know why I did it this way... :P